tests/integration/test_charm.py

```diff
@@ -78,6 +73,45 @@ async def test_prometheus_scrape_relation_with_prometheus_tester(
     tester_rules = get_rules_for(tester_app_name, rules_with_relation)
     assert len(tester_rules) == 1
 
+
+async def test_alert_rule_path_can_be_changed(ops_test, prometheus_tester_charm):
+    """Ensure scrape alert rules can be updated.
+
+    This test upgrades the metrics provider charm and checks that
+    updates to alert rules are propagated correctly.
+    """
+    # Change the alert rule path and ensure that there are 2
+    # after refreshing so they are reloaded
+    await ops_test.model.applications[tester_app_name].set_config(
+        {"alert-rules-path": "src/with_extra_alert_rule"}
+    )
+    await ops_test.model.wait_for_idle(apps=app_names, status="active")
+
+    await ops_test.model.applications[tester_app_name].refresh(
+        path=prometheus_tester_charm,
+        resources={
+            "prometheus-tester-image": oci_image(
+                "./tests/integration/prometheus-tester/metadata.yaml",
+                "prometheus-tester-image",
+            )
+        },
+    )
```

Comment:
This does not test alert rules changing across an upgrade because `set_config` -> `_configure` -> `_set_alerts` -> `push`, so the new alerts would be in effect before the upgrade.

Unfortuantely, `.refresh()` does not take a config dict. This was unexpected.

The Juju help CLI gives some hope:
```
Charm settings may be added or updated at upgrade time by specifying the
--config option, pointing to a YAML-encoded application config file.

  juju refresh foo --config config.yaml
```

Without this option in python-libjuju, I think this test is not giving us what we need.

Reply:
It does, actually. `_configure` is not involved in the hot path at all here. What we actually strike is `_set_scrape_job_spec` in `MetricsEndpointProvider` as part of `upgrade_charm`, and this invocation (after the upgrade) will have the alert rules from the config set, and the assertion after this verifies it.

It is _not_ testing whether upgrading Prometheus picks up alert rules from the config change (it won't, since the relation data will not be re-set unless the tester/provider is upgraded), but whether they can be changed at all now that the entire `write/remove/rebuild` pattern has been removed.